sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -324,14 +324,22 @@ object TypeCoercion {
       // We should cast all relative timestamp/date/string comparison into string comparisons
       // This behaves as a user would expect because timestamp strings sort lexicographically.
       // i.e. TimeStamp(2013-01-01 00:00 ...) < "2014" = true
-      case p @ BinaryComparison(left @ StringType(), right @ DateType()) =>
-        p.makeCopy(Array(left, Cast(right, StringType)))
-      case p @ BinaryComparison(left @ DateType(), right @ StringType()) =>
-        p.makeCopy(Array(Cast(left, StringType), right))
-      case p @ BinaryComparison(left @ StringType(), right @ TimestampType()) =>
-        p.makeCopy(Array(left, Cast(right, StringType)))
-      case p @ BinaryComparison(left @ TimestampType(), right @ StringType()) =>
-        p.makeCopy(Array(Cast(left, StringType), right))
+      // If StringType is foldable then we need to cast String to Date or Timestamp type
+      // which would give order of magnitude performance gain as well as preserve the behavior
+      // achieved by expressed above
+      // TimeStamp(2013-01-01 00:00 ...) < Cast( "2014" as timestamp) = true
+      case p @ BinaryComparison(left @ StringType(), right) if dateOrTimestampType(right) =>
+        if (left.foldable) {
+          p.makeCopy(Array(Cast(left, right.dataType), right))
```

Comment:
Thanks for pointing that out.
After applying the PR ` '10' < current_timestamp` does return `null` and its breaking the semantics.
The fix to this issue is very important. We have seen order of magnitude performance difference if string is correctly casted to timestamp for Literals in SQL and most of these SQLs are generated using some BI tool.
Any other suggestion to fix this issue? Shall i make it more restrictive so that null cases are better covered. i.e 
`if( expr.foldable && expressed.eval( null ) !=null ) 
  cast to timestamp`